// @slidewave/gitignore-include/.build/config.js
"use strict";
/* eslint-disable no-process-env */
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const Path = __importStar(require("path"));
class Config {
    constructor(options) {
        this.nodeEnv = options.nodeEnv;
        try {
            const packageJson = JSON.parse(fs_1.readFileSync(Path.resolve(__dirname, "..", "package.json"), {
                encoding: "utf8",
            }));
            if (packageJson) {
                this.npmConfig = packageJson;
            }
        }
        catch {
            // Don't care.
        }
    }
    get isProduction() {
        return this.nodeEnv === "production";
    }
}
exports.config = new Config({
    nodeEnv: process.env.NODE_ENV || "development",
});
